crates/turborepo-lib/src/package_manager.rs

```diff
@@ -134,13 +152,202 @@ impl PackageManager {
             exclusions,
         }))
     }
+
+    pub fn get_package_manager(base: &CommandBase, pkg: Option<&PackageJson>) -> Result<Self> {
+        // We don't surface errors for `read_package_manager` as we can fall back to
+        // `detect_package_manager`
+        if let Some(package_json) = pkg {
+            if let Ok(Some(package_manager)) = Self::read_package_manager(package_json) {
+                return Ok(package_manager);
+            }
+        }
+
+        Self::detect_package_manager(base)
+    }
+
+    fn detect_berry_or_yarn(version: &Version) -> Result<Self> {
+        let berry_constraint: Range = ">=2.0.0-0".parse()?;
+        if berry_constraint.satisfies(version) {
+            Ok(PackageManager::Berry)
+        } else {
+            Ok(PackageManager::Yarn)
+        }
+    }
+
+    fn detect_pnpm6_or_pnpm(version: &Version) -> Result<Self> {
+        let pnpm6_constraint: Range = "<7.0.0".parse()?;
+        if pnpm6_constraint.satisfies(version) {
+            Ok(PackageManager::Pnpm6)
+        } else {
+            Ok(PackageManager::Pnpm)
+        }
+    }
+
+    // Attempts to read the package manager from the package.json
+    fn read_package_manager(pkg: &PackageJson) -> Result<Option<Self>> {
+        let Some(package_manager) = &pkg.package_manager else {
+            return Ok(None)
+        };
+
+        let (manager, version) = Self::parse_package_manager_string(package_manager)?;
+        let version = version.parse()?;
+        let manager = match manager {
+            "npm" => Some(PackageManager::Npm),
+            "yarn" => Some(Self::detect_berry_or_yarn(&version)?),
+            "pnpm" => Some(Self::detect_pnpm6_or_pnpm(&version)?),
+            _ => None,
+        };
+
+        Ok(manager)
+    }
+
+    #[cfg(test)]
+    fn get_yarn_version(_: &AbsoluteSystemPathBuf) -> Result<Version> {
+        Ok(env::var("TEST_YARN_VERSION")?.parse()?)
+    }
+
+    #[cfg(not(test))]
+    fn get_yarn_version(project_directory: &AbsoluteSystemPathBuf) -> Result<Version> {
+        let output = Command::new("yarn")
+            .arg("--version")
+            .current_dir(&project_directory)
+            .output()?;
+        let yarn_version_output = String::from_utf8(output.stdout)?;
+        Ok(yarn_version_output.trim().parse()?)
+    }
```

Comment:
Maybe we introduce a `PackageManagerDetector` struct that holds a public `packageManager` field and a private `Option<Version>` where the getter initializes it and tests can just populate this field to avoid calling yarn.